### PR TITLE
[NOD-1210] Add integration test for IBD and fix bug where requestSelectedTipsIfRequired ran in handshake's goroutine

### PR DESCRIPTION
**Pull request by svarogg**

### CODE REVIEW

File: integration/main_test.go [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+import (
+	"os"
+	"testing"
+
+	"github.com/kaspanet/kaspad/logger"
+)
+
+func TestMain(m *testing.M) {
+	logger.SetLogLevels("trace") // TODO: Set back to debug
```
Comment (stasatdaglabs) [inline integration/main_test.go:11]: Please set this back to debug.

File: integration/notifications_test.go [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+import (
+	"testing"
+
+	"github.com/kaspanet/kaspad/wire"
+)
+
+func SetOnBlockAddedHandler(t *testing.T, harness *appHarness, handler func(header *wire.BlockHeader)) {
+	err := harness.rpcClient.NotifyBlocks()
+	if err != nil {
+		t.Fatalf("Error from NotifyBlocks: %+v", err)
```
Comment (stasatdaglabs) [inline integration/notifications_test.go:12]: Please use %s instead.

File: integration/setup_test.go [hunk 3]
```diff
@@ -14,41 +14,64 @@ type appHarness struct {
 	rpcClient       *rpcClient
 	p2pAddress      string
 	rpcAddress      string
+	minigAddress    string
+	minigAddressPK  string
```
Comment (stasatdaglabs) [inline integration/setup_test.go:18]: Typos: miningAddress, miningAddressPublicKey

File: integration/notifications_test.go [hunk 4]
```diff
@@ -0,0 +1,15 @@
+package integration
+
+import (
+	"testing"
+
+	"github.com/kaspanet/kaspad/wire"
+)
+
+func SetOnBlockAddedHandler(t *testing.T, harness *appHarness, handler func(header *wire.BlockHeader)) {
```
Comment (someone235) [inline integration/notifications_test.go:9]: Why is it exported?

File: integration/mining_test.go [hunk 5]
```diff
@@ -23,3 +25,24 @@ func solveBlock(t *testing.T, block *util.Block) *wire.MsgBlock {
 
 	panic("Failed to solve block! This should never happen")
 }
+
+func requestAndSolveTemplate(t *testing.T, harness *appHarness) *util.Block {
```
Comment (someone235) [inline integration/mining_test.go:29]: Because it also submits the block, I would rename it to "mineNextBlock"

**Pull request merged**